### PR TITLE
Declarative source base image

### LINKED ISSUE #22438: Declarative connectors runtime: Implement docker base image
The base docker image `airbyte/source-declarative-manifest` looks close to a regular low-code connector, but it has two key differences:
* Its manifest file is part of the ConnectionConfiguration, Instead of loading the manifest from a yaml file, it receives the config with the additional __injected_declarative_manifest and reads the manifest out of it before initializing the CDK base class with it.
* The ConnectorSpecification is incomplete. The spec for source-declarative-manifest only expects a “manifest” field, but it accepts additional properties. This means the connector’s spec operation will return an incomplete specification. To obtain the custom connector’s spec, the platform should instead read the spec from the actor_definition table. The spec in the actor_definition table will be populated by the connector builder, which knows about the additional expected fields.

The image contains the regular CDK code plus a thin wrapper for the initialization from the manifest. SourceLowcode acts as a wrapper around ManifestDeclarativeSource, and initializes it as part of the check, read, or discover operation because the manifest depends on the connection config.

```python
class GenericManifestDeclarativeSource(DeclarativeSource):
   def __init__(self):
       self._source = None

   @property
   def connection_checker(self) -> ConnectionChecker:
       return self._source.connection_check

   def streams(self, config):
       self._init_source(config)
       return self._source.streams(config)

   def spec(self, logger) -> ConnectorSpecification:
       raise Exception()

   def check(self, logger, config) -> AirbyteConnectionStatus:
       self._init_source(config)
       return self._source.check(logger, config)

   def read(self, logger, config, catalog, state):
       self._init_source(config)
       return self._source.read(logger, config, catalog, state)

   def _init_source(self, config):
       if self._source is None:
           self._source = ManifestDeclarativeSource(config['__injected_declarative_manifest'], True)
```

The (incomplete) spec can be hardcoded because every custom low-code connector expect the manifest to be present in the spec
The manifest field is airbyte_hidden because we don’t want the operator to configure the manifest. The platform will inject it.

Tests should be done on python level (not running connector acceptance tests).

If there is no passed in injected manifest as part of the config or the `__injected_declarative_manifest` property is not a valid low code manifest, the command will fail.

## Acceptance criteria

* `airbyte/source-declarative-manifest` is located next to the CDK code (in `airbyte-cdk/python`)
* There is a Dockerfile to build the `airbyte/source-declarative-manifest` image locally
* There are unit tests making sure it works as specified
* There is documentation explaining how this connector is meant to be used
* The spec command can be called but is expected to fail

### PR DESCRIPTION
**Pull request by flash1293**
## What

Fixes #22438

## How

WIP

## To test

Run `./gradlew airbyte-cdk:python:airbyteDocker`

Save `config.json`:
```
{
"__injected_declarative_manifest": {
  "version": "0.1.0",
  "definitions": {
    "selector": {
      "extractor": {
        "field_path": []
      }
    },
    "requester": {
      "url_base": "https://whiskyhunter.net/api",
      "http_method": "GET"
    },
    "retriever": {
      "record_selector": {
        "$ref": "#/definitions/selector"
      },
      "paginator": {
        "type": "NoPagination"
      },
      "requester": {
        "$ref": "#/definitions/requester"
      }
    },
    "base_stream": {
      "retriever": {
        "$ref": "#/definitions/retriever"
      }
    },
    "auctions_data_stream": {
      "$ref": "#/definitions/base_stream",
      "$parameters": {
        "name": "auctions_data",
        "path": "/auctions_data/?format=json"
      }
    },
    "auctions_info_stream": {
      "$ref": "#/definitions/base_stream",
      "$parameters": {
        "name": "auctions_info",
        "path": "/auctions_info?format=json"
      }
    },
    "distilleries_info_stream": {
      "$ref": "#/definitions/base_stream",
      "$parameters": {
        "name": "distilleries_info",
        "path": "/distilleries_info/?format=json"
      }
    }
  },
  "streams": [
    "#/definitions/auctions_data_stream",
    "#/definitions/auctions_info_stream",
    "#/definitions/distilleries_info_stream"
  ],
  "check": {
    "stream_names": [
      "auctions_data",
      "auctions_info",
      "distilleries_info"
    ]
  },
"spec": {
"type": "Spec",
  "documentation_url": "https://docs.airbyte.io/integrations/sources/whisky-hunter",
  "connection_specification": {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "title": "Whisky Hunter Spec",
    "type": "object",
    "additionalProperties": true,
    "properties": {}
  }
}

}
}
```

Save `catalog.json`:
```
{
"streams": [{"stream": {"name":"auctions_data", "json_schema": {}, "supported_sync_modes": ["full_refresh"]}, "sync_mode": "full_refresh", "destination_sync_mode": "overwrite"}]
}
```

Run `docker run --rm -v $(pwd)/config.json:/data/config.json -v $(pwd)/catalog.json:/data/catalog.json -i airbyte/source-declarative-manifest:dev read --config /data/config.json --catalog /data/catalog.json`

### CODE REVIEW

File: airbyte-cdk/python/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+# upgrade pip to the latest version
+RUN apk --no-cache upgrade \
+    && pip install --upgrade pip \
+    && apk --no-cache add tzdata build-base
+
+
+RUN mkdir cdk
+COPY ./setup.py ./cdk
+# FIXME install from pypi and make sure there are no race conditions
```
Comment (maxi297) [inline airbyte-cdk/python/Dockerfile:15]: FIXME

File: airbyte-cdk/python/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,36 @@
+FROM python:3.9.11-alpine3.15 as base
```
[the remark this answers is not in the document]
Comment (flash1293) [reply airbyte-cdk/python/Dockerfile:1]: Yes, but there is a separate issue for this: https://github.com/airbytehq/airbyte/issues/22442

**Review by flash1293 — COMMENTED** [review]
@maxi297  Tested with the most recent changes, still works fine.

I like the approach a lot, some things I'm not sure about:
* Is the Dockerfile right in the `airbyte-cdk/python` directory the right approach or should we move it into a subfolder with it's own `build.gradle` file?
* Should we implement a `Dockerfile.test` (supported easily via the airbyte docker gradle 
  plugin) which is using the local cdk version for easy testing instead of the released version? Doesn't seem like a huge lift - if you have a `Dockerfile.test`, then there is a gradle task `:airbyteDockerTest` which builds the image using that Dockerfile so we can quickly test whether things work
* Can/should we have unit tests for the `main.py`? Not a hard requirement IMHO
* I assume we are going to implement the autobump for the `RUN pip install --prefix=/install airbyte-cdk==0.29.0` line in the Dockerfile as part of https://github.com/airbytehq/airbyte/issues/22442, right?

**Comment by maxi297** [comment]
> Is the Dockerfile right in the airbyte-cdk/python directory the right approach or should we move it into a subfolder with it's own build.gradle file?

If we have reasons to believe it is not, this would be useful input to understand where it needs to be located. Personally, I don't mind not taking a decision now since it seems very easy to change once we decide it needs to be change for whatever reason.

> Should we implement a Dockerfile.test (supported easily via the airbyte docker gradle
plugin) which is using the local cdk version for easy testing instead of the released version? Doesn't seem like a huge lift - if you have a Dockerfile.test, then there is a gradle task :airbyteDockerTest which builds the image using that Dockerfile so we can quickly test whether things work

Do we need this? If I want to do local testing, it seems simplier (at least for me as a developper) to call `python main.py <comment> <args>`. Since this is simple and doesn't require to maintain another Dockerfile, I personally prefer this but I'm willing to change this if you see the need to.

> Can/should we have unit tests for the main.py? Not a hard requirement IMHO

It makes sense. I'll do that

> I assume we are going to implement the autobump for the RUN pip install --prefix=/install airbyte-cdk==0.29.0 line in the Dockerfile as part of https://github.com/airbytehq/airbyte/issues/22442, right?

autobump is already implemented (see https://github.com/airbytehq/airbyte/pull/23275/files#diff-d6bc49914d963ed4d9a2df5994d349fe0ee6ebe5649ba430e2f4dc983c57ce11R6)

**Comment by maxi297** [comment]
> I thought this would only take care of the label, not the pip install command, but if that just works out of the box, that’s great.

I'm not sure I get it. So what autobump does is update the version defined in `.bumpversion.cfg` in all the specified files. In this PR, autobump will check in `setup.py` and `Dockerfile` for the version and update it accordingly. What will happen on release:
* autobump to bump versions in setup.py and the Dockerfile (for line `RUN pip install --prefix=/install airbyte-cdk==<version>` and `LABEL io.airbyte.version=<version>`
* push new airbyte_cdk to pypi using <new version>
* build/publish docker image using `RUN pip install --prefix=/install airbyte-cdk==<new version>` and `LABEL io.airbyte.version=<new version>`

I think there might be some work to be done on the publishing (to be honest, I don't know a lot about our current /publish but I guess this is the work that needs to be figured out as part of https://github.com/airbytehq/airbyte/issues/22442. To be honest, I added autobump to make sure the versions would stay in sync even though we don't publish yet. 

If there is something I don't see, let me know and we'll figure it out together

**Comment by flash1293** [comment]
@maxi297 I think we are good, I thought it would be some special thing looking for the label but if it's a search/replace algorithm that's great - just a misunderstanding from my side, nothing to do here.

**Review by maxi297 — APPROVED** [review]
Approving for Joe since he started the PR and therefore can't do it

File: airbyte-cdk/python/Dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (girarda) [reply airbyte-cdk/python/Dockerfile:1]: cool! Approved!

**Review by girarda — APPROVED** [review]

**Pull request merged**